pkg/ctl/install/pki.go

```diff
@@ -0,0 +1,167 @@
+)
+
+type publicKeyInfrastructurePaths struct {
+	caCertificate     string
+	caKey             string
+	serverCertificate string
+	serverKey         string
+	clientCertificate string
+	clientKey         string
+}
+
+type publicKeyInfrastructure struct {
+	caCertificate     []byte
+	caKey             []byte
+	serverCertificate []byte
+	serverKey         []byte
+	clientCertificate []byte
+	clientKey         []byte
+}
+
+type devnull struct{}
+
+func (devnull) Debug(string)   {}
+func (devnull) Info(string)    {}
+func (devnull) Warning(string) {}
+func (devnull) Err(string)     {}
+func (devnull) Crit(string)    {}
+func (devnull) Emerg(string)   {}
+func init() {
+	// Disable logging of cfssl since we handle errors explicitly
+	log.SetLogger(devnull{})
+}
+
+func newPKI(hostName string, validFor time.Duration, rsaKeyBitSize int) (*publicKeyInfrastructure, error) {
+	keyReq := &csr.KeyRequest{
+		A: "rsa",
+		S: rsaKeyBitSize,
+	}
+
+	// Generate CA
+	caReq := &csr.CertificateRequest{
+		KeyRequest: keyReq,
+		CN:         "CA",
+	}
+	caCert, _, caKey, err := initca.New(caReq)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate root CA: %s", err)
+	}
+
+	// Generate Server Certificate
+	serverCert, serverKey, err := generateCertificate(caCert, caKey, keyReq, hostName, "server", validFor)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate server certificate: %s", err)
+	}
+
+	// Generate Client certificate
+	generateCertificate(caCert, caKey, keyReq, hostName, "client", validFor)
+	clientCert, clientKey, err := generateCertificate(caCert, caKey, keyReq, hostName, "client", validFor)
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate client certificate: %s", err)
+	}
+	pki := &publicKeyInfrastructure{
+		caCertificate:     caCert,
+		caKey:             caKey,
+		serverCertificate: serverCert,
+		serverKey:         serverKey,
+		clientCertificate: clientCert,
+		clientKey:         clientKey,
+	}
+	return pki, nil
+
+}
+
+func generateCertificate(caCert []byte, caKey []byte, keyReq *csr.KeyRequest,
+	hostName string, commonName string, validFor time.Duration) ([]byte, []byte, error) {
+	policy := &config.Signing{
+		Default: &config.SigningProfile{
+			Expiry: validFor,
+			Usage:  []string{"signing", "key encipherment", "server auth", "client auth"},
+		},
+	}
+	parsedCa, err := helpers.ParseCertificatePEM(caCert)
+	if err != nil {
+		return nil, nil, fmt.Errorf("malformed generated private CA certificate: %s", err)
+	}
+	priv, err := helpers.ParsePrivateKeyPEM(caKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("malformed generated private CA key: %s", err)
+	}
+	certSigner, err := local.NewSigner(priv, parsedCa, signer.DefaultSigAlgo(priv), policy)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot create signer: %s", err)
+	}
+	serverReq := &csr.CertificateRequest{
+		KeyRequest: keyReq,
+		CN:         commonName,
+		Hosts:      []string{hostName},
+	}
+	g := &csr.Generator{Validator: genkey.Validator}
```

Comment:
From the docs, my first guess would be  that `csr.Generate` is what you wan to call?

Reply:
I does seem all a little verbose, are you sure there's in a higher-level option?

Reply:
I am basing this code on what the command line tool does. If you think there is a better way, please suggest it.

Reply:
I will try with `csr.Generate`

Reply:
I simplified it